### PR TITLE
feat(#197): gbrain memory backend + CRDB adapter + hybrid composer (default)

### LINKED ISSUE #197: Capability loop #T: @skytwin/memory-gbrain + memory-hybrid composer + CRDB adapter (target v1.0.5)
**Parent epic:** #195
**Phase:** 4 — Ecosystem (target ship: v1.0.5, ~2 weeks post-OSS-launch)
**Effort:** ~55h (was 40h; CRDB adapter scope added — see DB stack note below)
**Depends on:** #196 (`@skytwin/memory-port` is the contract this child implements)

## Context

The first proof of the memory-swap thesis. Adds a gbrain-backed implementation of `MemoryPort` and a hybrid composer that routes operations to whichever backend serves them best.

**DB-stack reality:** gbrain ships defaulting to PGLite or Supabase. SkyTwin's primary DB stack is **CockroachDB** (default, multi-user) and **SQLite** (single-user turnkey mode via #188). Running gbrain on its own separate PGLite means two databases per install — undesirable operationally. This child includes a **CockroachDB adapter for gbrain** so installations that already run CRDB don't need a second Postgres. CRDB has native VECTOR type (post-2024) + Postgres-wire-compatible tsvector; gbrain's pgvector + tsvector usage translates with a small driver-level shim. RRF query layer works as-is. Durable job queue ("minions") benefits from CRDB serializable transactions.

The SQLite-vec adapter (for #188 single-user turnkey installs) is **not** in this child's scope — it's a fast-follow for v1.1. Until both ship, gbrain in single-user mode falls back to bundled PGLite (acceptable v1.0.5 trade-off).

**Why post-launch (v1.0.5), not at launch:** gbrain isn't installed on the launch machine; its install flow needs validation; coupling SkyTwin's launch to gbrain readiness is launch risk for an architectural commitment that the *port itself* (#196) already proves. Shipping the port at launch and gbrain two weeks later turns one launch beat into two press cycles ("memory is a port" → "first swap, your twin just got 30% faster at semantic search").

## Scope

- New package `@skytwin/memory-gbrain` — implements `MemoryPort` against gbrain's MCP interface.
  - Translates `MemoryPort` operations into gbrain CLI/MCP calls (`gbrain search`, `gbrain query`, `gbrain code-def`, etc.).
  - Declares capabilities `{semantic_search, code_aware_search, federated_sources}`.
  - Polyfills via `SignalsRouter` for the operations gbrain doesn't natively support (graph walks, episodic, spatial wings, AAAK).
- New package `@skytwin/memory-gbrain-crdb-adapter` — driver-level shim that lets gbrain run against CockroachDB instead of bundled PGLite.
  - Translates pgvector calls to CRDB's native VECTOR primitives.
  - Translates tsvector queries (Postgres-wire-compatible on CRDB; mostly pass-through with edge-case handling for any tsvector functions CRDB hasn't reached parity on yet).
  - RRF query layer used as-is (works on either backend).
  - Job queue ("minions") uses CRDB serializable transactions natively — small adapter to map gbrain's job queue table semantics.
  - Verified against the CRDB version SkyTwin pins (current production target).
- New package `@skytwin/memory-hybrid` — composes any two `MemoryPort` implementations.
  - Routes each operation to whichever backend declares the relevant capability natively (preferring native > polyfill).
  - Default routing: gbrain serves `searchSemantic` and `code_aware_search`; MemPalace serves `walkGraph`, `getEpisodes`, `getTriples`, `summarize`, `compress`.
  - Writes go to both backends; reads route per-operation.
  - Configurable routing rules per-installation.
- gbrain install/setup is delegated to the `setup-gbrain` skill flow (already exists in gstack); SkyTwin auto-detects gbrain on `~/.config/gbrain/` and offers to enable hybrid mode.
- "Your twin just got smarter" notification UX when the hybrid mode activates.

## Acceptance Criteria

1. Package `@skytwin/memory-gbrain` builds clean; implements `MemoryPort`; talks to gbrain via its MCP interface (using `@skytwin/mcp-host` from #173).
2. Capabilities declared accurately: `{semantic_search, code_aware_search, federated_sources}`.
3. Polyfills route correctly via `SignalsRouter`: a `walkGraph` request against `memory-gbrain` alone succeeds (slowly) via repeated `getTriples` synthesized from gbrain's semantic search results.
4. Package `@skytwin/memory-hybrid` builds clean; takes two `MemoryPort` instances + routing rules; routes correctly per `port.capabilities()`.
5. Default hybrid configuration: gbrain for semantic + code-aware; MemPalace for graph + episodic + spatial. Verified by integration test.
6. Writes go to both backends in hybrid mode; consistency verified by round-trip test.
7. gbrain detection: SkyTwin checks for `~/.config/gbrain/` on startup; if present, surfaces "enable hybrid mode" prompt; opt-in.
8. "Your twin just got smarter" notification fires once on first hybrid-mode activation; user dismissable; provenance node written.
9. Performance: hybrid-routed `searchSemantic` is ≥30% faster than MemPalace-only on a fixture corpus of 10k user signals.
10. Migration support: existing MemPalace data accessible from hybrid mode without import; new gbrain-side index built lazily on first hybrid query.
11. Tests: ≥10 unit (gbrain wrapper contracts), ≥5 integration (real gbrain instance against fixture data), ≥3 E2E (full SkyTwin loop in hybrid mode).
12. **CRDB adapter**: gbrain runs end-to-end against a CockroachDB instance (no separate Postgres process); pgvector → CRDB VECTOR translation verified; tsvector queries work; RRF retrieval results match the equivalent PGLite-backed run within ±2% on the fixture corpus.
13. **CRDB adapter tests**: ≥6 unit (vector type translation, tsvector edge cases, job queue semantics), ≥3 integration (gbrain-on-CRDB against the SkyTwin DB stack).
14. SQLite single-user installs (#188) gracefully fall back to bundled PGLite for gbrain in v1.0.5 (with surfaced log/UI note "gbrain on SQLite-vec is a v1.1 fast-follow").

## Out of Scope (this child)

- Future memory engines as they land. When they ship, they become additional implementations; pattern from this child applies.
- gbrain authoring or contributing back upstream. We *consume* gbrain.
- Migration tooling between backends beyond `port.exportAll()` / `port.importAll()` from #196.

## Files Reference

| File | Change |
|---|---|
| `packages/memory-gbrain/` | New package |
| `packages/memory-hybrid/` | New package |
| `packages/memory-gbrain/src/index.ts` | gbrain MCP wrapper implementing MemoryPort |
| `packages/memory-hybrid/src/router.ts` | Hybrid routing logic |
| `apps/api/src/routes/memory-config.ts` | New: enable/disable hybrid; show backend status |
| `apps/web/public/js/pages/memory-settings.js` | New: backend selection UI |
| `docs/memory-swap.md` | New: how to swap memory backends |

## Rollback

Hybrid mode is opt-in. If gbrain misbehaves, set `MEMORY_BACKEND=mempalace-only` in config; system reverts to single-backend MemPalace operation.

## Why this validates the thesis

This child is the *proof* that memory-port-as-architecture works. Anyone can swap memory backends without touching consumers. Whatever memory engine ships next, when it lands, follows the same pattern. The hybrid composer means users don't have to choose — they get the union of capabilities.

Most agent products in 2026 will hardcode their memory. SkyTwin's commitment to portability is a differentiator that compounds over time as memory architectures continue to mutate.

### PR DESCRIPTION
**Pull request by jayzalowitz**
Closes #197.

Promotes `@skytwin/memory-gbrain` from a CLI-shellout skeleton (PR #215) to a real, in-process, CockroachDB-backed memory layer. The default `MemoryPort` for new installs is now gbrain — vector embeddings + tsvector full-text search, fused via Reciprocal Rank Fusion. **No separate Postgres process, no external CLI install** — gbrain runs against the SkyTwin DB stack directly.

User direction was explicit: gbrain is the **default**, mempalace is the **second option**, and everything works against CRDB where possible. Done.

### New package: `@skytwin/memory-gbrain-crdb-adapter`
- `repository.ts` — CRDB-backed CRUD + `hybridSearch` (parallel text + vector, application-side RRF fold). Per-user `pendingEmbeddingJobs(userId?)`.
- `in-memory-repository.ts` — same surface, in-process; used by tests.
- `embedding.ts` — `EmbeddingProvider` interface + `HashEmbeddingProvider` (deterministic, hash-trick, zero-config) and `OpenAiEmbeddingProvider` (OpenAI-compatible — works with Ollama, llamafile, vLLM).
- `rrf.ts` — Reciprocal Rank Fusion fold (k=60).

### Migration `040-gbrain-memory.sql`
- `brain_pages` (FLOAT8[] + TSVECTOR + INVERTED INDEX), `brain_entities`, `brain_triples`, `brain_episodes`, `brain_signals`, `brain_settings`, `brain_embedding_jobs` (durable queue with `SELECT FOR UPDATE SKIP LOCKED`).
- **`id` is `STRING PRIMARY KEY DEFAULT gen_random_uuid()::STRING`** (not UUID) — production signal IDs like `sig_gmail_abc` need to round-trip.

### Updated `@skytwin/memory-gbrain`
- `EmbeddedGbrainMemoryPort` — full `MemoryPort` implementation. Capabilities: `semantic_search`, `code_aware_search`, `temporal_triples`, `episodic`, `graph_walk`. Synchronous embedding when fast, async via the job queue when an external provider is configured. `candidatePoolSize` defaults to `max(k*4, 40)` per query (was pinned 40).
- `searchCodeAware` — 1.25× boost for `source = 'code'` pages.
- `hasExternalGbrainConfig()` — detects existing `~/.config/gbrain/` for the dashboard prompt.

### Updated `@skytwin/memory-hybrid`
- `HybridDiagnostics` counters expose routing + write outcomes.
- `resolveReadPort` falls through to secondary when primary lacks the relevant capability.
- `getEntitiesByType` now routes through `resolveReadPort` (was hard-wired secondary).

### API + Web
- `apps/api/src/memory-setup.ts` — per-user backend factory. Default `gbrain`; `MEMORY_BACKEND` env override; per-user `brain_settings.backend` wins. The `mempalace` and `hybrid`-secondary paths now wire a **real** `MemPalaceMemoryPort` against `mempalaceRepository` (previously a stub that returned `[]`).
- `apps/api/src/routes/memory-config.ts` — `GET/POST /api/memory-config`, `POST /dismiss-notification`, `GET /diagnostics`, `GET /dashboard`.
- `apps/web/public/js/pages/memory-settings.js` — backend switcher + capability list + "Your twin just got smarter" notice + memory dashboard (recent episodes, top entities, feedback histogram).

### Decision pipeline wiring
- `SenderAwareCandidateGenerator` — closes the rule-based safety gap surfaced by the fake-user E2E: board / CFO / legal / investor senders get a `flag_for_manual_review` candidate that suppresses the rule-based `archive_email` set entirely. The built-in `NO_IRREVERSIBLE_WITHOUT_APPROVAL` policy gates it to the approval queue at every trust tier.
- `events.ts` — wires this generator as both the standalone rule-based DecisionMaker and the inner `RuleBasedCandidateGenerator` of the LLM fallback chain. Populates `DecisionContext.episodicMemories` from `mempalaceRepository.getEpisodes` so the existing `calculateEpisodicBoost` actually fires. Best-effort writes the inbound signal into `MemoryPort.recordSignal`.
- `approvals.ts` — on approve/reject, records an episode into both the legacy table AND the gbrain MemoryPort.
- `assistant.ts` — chat context provider queries `MemoryPort.searchSemantic` in parallel with mempalace's ILIKE; results deduped.

### Embedding worker
- `apps/worker/src/jobs/embedding-backfill.ts` — drains `brain_embedding_jobs` every 30s. SELECT FOR UPDATE SKIP LOCKED makes it safe under multiple worker instances. Env-driven provider selection mirrors `apps/api/src/memory-setup.ts`.

### Documentation
- `docs/memory-swap.md` — full runbook, capability table, env knobs, rollback.
- `docs/architecture-philosophy.md` — memory port row updated.
- `docs/cockroach-architecture.md` — 7 brain_* tables documented + the STRING-PK rationale.
- `docs/technical-spec.md` — 5 new memory-* packages in the layout + topological dependency chain.
- `CHANGELOG.md` — substantial unreleased entry.
- `CLAUDE.md` — package table refreshed.

## Tests

| Package | Tests |
|---|---|
| `@skytwin/memory-gbrain` | 86 |
| `@skytwin/memory-gbrain-crdb-adapter` | 50 + 6 DB-gated |
| `@skytwin/memory-hybrid` | 19 |
| `@skytwin/decision-engine` | +17 new (sender-aware 12, corrections 5) |
| `@skytwin/worker` | +12 new (embedding-backfill) |
| `@skytwin/api` | +48 new (config 8, roundtrip 5, setup 13, events-memory 3, fake-user 16, feedback-loop 4, full-loop 1, gbrain-write 1, dashboard 5) |
| **Total** | **~245 new tests** |

Full suite: **70/70 turbo tasks pass** (api 542, decision-engine 109, memory-gbrain 86, memory-gbrain-crdb-adapter 50, memory-hybrid 19, worker 83).

## Post-/review fixes (Copilot)

**Round 1 (commit 42283af):**

| # | Finding | Fix |
|---|---|---|
| 1 | UUID PK on brain_* tables would reject non-UUID signal ids | STRING PK with UUID-string default |
| 2 | `StubMempalacePort` returned `[]` everywhere | Real `MemPalaceMemoryPort` wired to `mempalaceRepository` |
| 3 | `pendingEmbeddingJobs` global, not per-user | Optional `userId`; dashboard passes it |
| 4 | `candidatePoolSize` hard-coded 40 vs doc's `max(k*4, 40)` | Per-query computation |
| 5 | In-memory `embeddingModel` mismatch with CRDB path | Conditional spread |
| 6 | `event.target.closest` could throw on text-node clicks | `instanceof Element` guard |
| 7 | `getEntitiesByType` bypassed `resolveReadPort` | Added to RoutingRules; routes properly |
| 8 | docs claimed mempalace had no `semantic_search` | Table now shows ILIKE + note |

**Round 2 (commit 22e5689) — Copilot re-reviewed on 50f297c:**

| # | Finding | Fix |
|---|---|---|
| 1 | `brain_embedding_jobs.page_id UUID` vs `brain_pages.id STRING` FK type mismatch — inserts of non-UUID page ids would fail | Migration 040: `page_id STRING` matches `brain_pages.id` |
| 2 | assistant.ts comment said dedupe by `(summary, occurredAt)` but code dedupes by summary only | Comment corrected — including occurredAt would defeat cross-source dedupe (gbrain hits never have one) |
| 3 | hybrid-port.ts `resolveReadPort` docstring claimed 3-step priority but code has 2 | Docstring rewritten to match single-rule + capability-fallback logic |
| 4 | `.claude/scheduled_tasks.lock` (volatile session lock) leaked into the PR | `git rm --cached` + added to .gitignore |
| 5 | `BrainPageRow.embedding: number[] \| null` mismatched parsePageRow's `typeof === 'string'` runtime check (strict-mode always-false flag) | New `RawBrainPageRow` type with `number[] \| string \| null` for the raw row; `parsePageRow` narrows to `BrainPageRow` |

## Operational notes

- `MEMORY_BACKEND=gbrain` (default) | `hybrid` | `mempalace` switches the per-install default; per-user override via dashboard.
- `OPENAI_EMBEDDING_API_KEY` (or `OPENAI_API_KEY`) → real embeddings. `OPENAI_EMBEDDING_BASE_URL` points at any OpenAI-compatible endpoint (Ollama, llamafile, vLLM, LocalAI). See [`docs/memory-swap.md`](../blob/jayzalowitz/gbrain-default-impl/docs/memory-swap.md) for the Ollama recipe.
- Rollback: `MEMORY_BACKEND=mempalace`. Legacy `/api/mempalace` routes unaffected — they query `memory_*` tables directly, not via MemoryPort.

## Polish (commits aa78f13 → 50f297c)

After Copilot review fixes landed, three follow-up items shipped on top:

- **SSE live memory dashboard** — `events.ts` emits `memory:page-indexed` after `recordSignalToMemory`; `approvals.ts` emits `memory:episode-recorded` after `createEpisode`. Web client subscribes and re-renders the dashboard with a 1s debounce. No polling. Unit-tested in `feedback-loop.test.ts` and `gbrain-write-on-events.test.ts`.
- **Ollama recipe** — `docs/memory-swap.md` now documents a zero-cloud local-embeddings setup via Ollama's OpenAI-compatible `/v1/embeddings` endpoint (`nomic-embed-text` by default).
- **CRDB integration harness** — `packages/memory-gbrain-crdb-adapter/scripts/run-crdb-integration.sh` + a `test:crdb` package script. Spins a hermetic Docker CRDB, applies migration 040, seeds a test user, runs the 6 DB-gated tests, tears down on exit. Post-/review pass hardened the script: `ON_ERROR_STOP=on` on every psql call, ready-flag wait loop with container logs on timeout, robust TEST_USER_ID parsing.

## Documentation

`/document-release` ran post-merge-gate against the final pre-merge state:

| File | Status |
|---|---|
| `README.md` | Current — memory-backend bullet already updated |
| `CLAUDE.md` | Current — new memory packages listed |
| `CHANGELOG.md` | **Fixed** — restored the "Embedded LLM round-3" `[unreleased]` entry that an earlier rebase silently dropped (PR #249 commit `c6e93de`); squash-merge would have deleted it from main. See commit 50f297c. |
| `docs/architecture-philosophy.md` | Current (afc5b5c) |
| `docs/cockroach-architecture.md` | Current (afc5b5c) — brain_* tables + STRING PK rationale |
| `docs/technical-spec.md` | Current (afc5b5c) — 5 new memory-* packages in layout |
| `docs/memory-swap.md` | Current (aa78f13 + d112aaa) — Ollama recipe, CRDB harness, prose aligned with capability table |
| `TODOS.md` | Current — no memory-related items to close |
| `VERSION` | Skipped — user decides at ship time |

## Test plan

- [x] `pnpm build --concurrency=1` → 35/35 packages
- [x] `pnpm test` → 70/70 turbo tasks (544 API tests pass, 24 skipped — DB-gated)
- [x] Rebased onto main
- [x] All 7 Copilot inline + 1 summary finding addressed (with inline replies pointing at fix commits)
- [x] `/review` re-run on post-Copilot diffs → only `run-crdb-integration.sh` issues found; auto-fixed in aa93c33
- [x] `/document-release` re-run → CHANGELOG drift caught and restored in 50f297c
- [x] Copilot re-reviewed 50f297c → 5 round-2 findings; all addressed in 22e5689 with inline replies pointing at fixes
- [ ] CI green on 22e5689
- [ ] Manual smoke against a live CRDB (optional; `pnpm --filter @skytwin/memory-gbrain-crdb-adapter test:crdb`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by jayzalowitz** [comment]
## All Copilot findings addressed (commit `42283af`)

Rebased onto main and addressed every inline finding plus the summary comment:

| # | Finding | Fix |
|---|---|---|
| 1 | `brain_signals.id UUID PRIMARY KEY` rejects non-UUID signal ids | Changed brain_pages/entities/triples/episodes/signals id to `STRING PRIMARY KEY DEFAULT gen_random_uuid()::STRING`. brain_settings.user_id stays UUID (real FK); brain_embedding_jobs.id stays UUID (internal-only). |
| 2 | `StubMempalacePort` returned `[]` everywhere; hybrid secondary lost data | Real `MemPalaceMemoryPort` wired with a full `MemPalaceRepos` adapter against `mempalaceRepository`. Covers knowledgeGraph + episode surfaces. Palace/closet/entityCode throw (unreachable via MemoryPort). |
| 3 | `pendingEmbeddingJobs` counted globally on multi-user installs | Optional `userId` parameter. Dashboard passes userId; worker passes none (global queue depth is the right number for drain telemetry). |
| 4 | `candidatePoolSize` doc promised `max(k*4, 40)` but constructor pinned 40 | Store override as sentinel; compute `max(k*4, 40)` per query in `searchInternal`. |
| 5 | In-memory `embeddingModel` set even when embedding failed (divergent from CRDB path) | Match CRDB path: only spread `embeddingModel` when embedding succeeded. Applied to recordSignal/recordEntity/recordEpisode. |
| 6 | `event.target.closest` could throw on text-node clicks | Added `instanceof Element` guard per CLAUDE.md frontend event-handling rule. |
| 7 | `getEntitiesByType` bypassed `resolveReadPort` | Added `getEntitiesByType` to `RoutingRules` (defaults to `primary`); route through `resolveReadPort` with `temporal_triples` capability check. |
| 8 | docs/memory-swap.md said mempalace had no `semantic_search` | Updated capability table to show `ILIKE` with a note explaining the keyword-vs-semantic distinction. |

Also resolved the merge conflict with main (#248 first-run dashboard merged between this PR's commits). Both CHANGELOG entries are stacked under unreleased.

Full suite: 70/70 turbo tasks pass.

File: packages/memory-gbrain/src/embedded-port.ts [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+      throw new Error('EmbeddedGbrainMemoryPort: userId is required');
+    }
+    this.userId = opts.userId;
+    this.embedding = opts.embedding ?? new HashEmbeddingProvider();
+    this.backend = opts.backend ?? 'crdb';
+    this.store = opts.store ?? (this.backend === 'memory' ? new InMemoryBrainStore() : undefined);
+    this.rrfK = opts.rrfK ?? 60;
+    this.candidatePoolSize = opts.candidatePoolSize ?? 40;
+    this.embedQueriesSync = opts.embedQueriesSynchronously ?? true;
+  }
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply packages/memory-gbrain/src/embedded-port.ts:113]: Addressed in 42283af (post-/review fixes): `candidatePoolSize` is now `candidatePoolOverride` (stored only when explicitly set) and the search path computes `Math.max(k*4, 40)` per query at embedded-port.ts:366. Reading a large `k` no longer caps the candidate pool at 40.

File: packages/memory-gbrain/src/embedded-port.ts [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+      this.store!.insertPage({
+        userId: this.userId,
+        title: `${s.source}/${s.type}`,
+        content: summaryText,
+        source: 'signal',
+        sourceRef: s.id,
+        metadata: { signalSource: s.source, signalType: s.type },
+        embedding: await this.embedding.embed(summaryText).catch(() => undefined),
+        embeddingModel: this.embedding.model,
+      });
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply packages/memory-gbrain/src/embedded-port.ts:162]: Addressed in 42283af. The in-memory `recordSignal`/`recordEntity`/`recordEpisode` paths now use the same conditional `embeddingModel` spread as the CRDB paths — `{ ...(embedding ? { embedding, embeddingModel: this.embedding.model } : {}) }` — so pages with a failed embedding never get a non-null model. See embedded-port.ts:161.

File: apps/api/src/routes/memory-config.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      return res.status(400).json({ error: 'invalid userId' });
+    }
+    try {
+      const [settings, resolved, suggestion, counts, pendingJobs] = await Promise.all([
+        getBrainSettings(userId).catch(() => null),
+        getMemoryPortForUser(userId),
+        Promise.resolve(suggestHybridUpgrade()),
+        countPages(userId).catch(() => ({ total: 0, embedded: 0 })),
+        pendingEmbeddingJobs().catch(() => 0),
+      ]);
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply apps/api/src/routes/memory-config.ts:51]: Addressed in 42283af. `pendingEmbeddingJobs` now accepts an optional `userId`: the API/dashboard routes pass it (memory-config.ts:50, :160), and the worker keeps the global count for its drain-loop telemetry. See repository.ts:537.

File: apps/api/src/memory-setup.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  capabilities(): Set<MemoryCapability> {
+    return new Set<MemoryCapability>(['spatial_wings', 'aaak_compression']);
+  }
+  async recordSignal() {}
+  async recordEntity() {}
+  async recordTriple() {}
+  async recordEpisode() {}
+  async searchSemantic() { return []; }
+  async walkGraph() { return []; }
+  async getEpisodes() { return []; }
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply apps/api/src/memory-setup.ts]: Addressed in 42283af. `buildMempalacePort()` now returns a real `MemPalaceMemoryPort` wired to `mempalaceRepository` (memory-setup.ts:110-232), used both for `MEMORY_BACKEND=mempalace` and as the hybrid secondary. The stub is gone.

File: apps/web/public/js/pages/memory-settings.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  });
+}
+
+function ensurePageListener() {
+  if (_pageListenerWired) return;
+  _pageListenerWired = true;
+
+  document.addEventListener('click', async (event) => {
+    if (window.location.hash.split('?')[0] !== '#/memory-settings') return;
+    const target = event.target.closest('[data-action]');
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply apps/web/public/js/pages/memory-settings.js:74]: Addressed in 42283af. The delegated handler now has `if (!(event.target instanceof Element)) return;` before `event.target.closest(...)` (memory-settings.js:73-74), matching the pattern used elsewhere in the SPA.

File: docs/memory-swap.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+  most retrieval falls back to keyword search.
+
+## Backends at a glance
+
+| backend     | semantic_search | code_aware_search | episodic | graph_walk | temporal_triples | spatial_wings | aaak_compression |
+|-------------|:---------------:|:-----------------:|:--------:|:----------:|:----------------:|:-------------:|:----------------:|
+| `gbrain`    |        ✓        |         ✓         |    ✓     |     ✓      |        ✓         |               |                  |
+| `hybrid`    |        ✓        |         ✓         |    ✓     |     ✓      |        ✓         |       ✓       |        ✓         |
+| `mempalace` |                 |                   |          |            |                  |       ✓       |        ✓         |
+
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply docs/memory-swap.md:36]: Addressed in 42283af (table) and d112aaa (prose). The capability table now shows `mempalace` as ILIKE-backed for `semantic_search`, and the bullet at the top of memory-swap.md was tightened to match — it had still said "declares no semantic_search capability" which contradicted both the code and the table. `MemPalaceMemoryPort.capabilities()` does declare `semantic_search`; the runbook now explains it's keyword-ILIKE, not vector.

File: packages/db/src/migrations/040-gbrain-memory.sql [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+-- shape so MemoryRecord round-trips without information loss.
+CREATE TABLE IF NOT EXISTS brain_signals (
+  id UUID PRIMARY KEY,
+  user_id UUID NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+  source STRING NOT NULL,
+  type STRING NOT NULL,
+  data JSONB NOT NULL DEFAULT '{}',
+  recorded_at TIMESTAMPTZ NOT NULL DEFAULT now(),
+  signal_timestamp TIMESTAMPTZ NOT NULL
+);
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply packages/db/src/migrations/040-gbrain-memory.sql:137]: Addressed in 42283af. Migration 040 now uses `id STRING PRIMARY KEY DEFAULT gen_random_uuid()::STRING` on `brain_pages`, `brain_entities`, `brain_triples`, `brain_episodes`, and `brain_signals` so non-UUID source IDs (`sig_gmail_abc123`, `msg-1`) are accepted directly without an extra surrogate column. The default still generates UUID-shaped strings when the caller doesn't supply an id.

File: packages/db/src/migrations/040-gbrain-memory.sql [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+-- ============================================================================
+-- Embeddings are computed asynchronously so write paths (recordSignal,
+-- recordEntity) stay fast. CRDB's serializable transactions give us
+-- "at-most-once" claim semantics for free — a worker SELECT...FOR UPDATE
+-- SKIP LOCKED leases a job; the lease auto-expires if the worker dies.
+CREATE TABLE IF NOT EXISTS brain_embedding_jobs (
+  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+  user_id UUID NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+  page_id UUID NOT NULL REFERENCES brain_pages(id) ON DELETE CASCADE,
+  status STRING NOT NULL DEFAULT 'pending'
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply packages/db/src/migrations/040-gbrain-memory.sql:170]: Addressed in 22e5689. `brain_embedding_jobs.page_id` is now `STRING` to match `brain_pages.id` exactly; the FK and inserts with non-UUID page ids both work now. Migration 040 line ~169 carries an inline comment explaining the earlier-draft history.

File: apps/api/src/routes/assistant.ts [hunk 9]
```diff
@@ -131,30 +132,77 @@ function buildContextBuilder(): ContextBuilder {
...
-        // human-readable hint, not a full structured payload.
         outcome: r.outcome ? renderOutcomeHint(r.outcome) : undefined,
         occurredAt: r.created_at instanceof Date ? r.created_at.toISOString() : undefined,
       }));
+
+      // Dedupe by summary text — same episode may surface from both sources.
+      const seen = new Set<string>();
+      const merged: MergedHit[] = [];
+      for (const item of [...fromSemantic, ...fromMempalace]) {
+        const key = item.summary.trim().toLowerCase();
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply apps/api/src/routes/assistant.ts:202]: Addressed in 22e5689 by fixing the comment, not the code. gbrain hits never carry an `occurredAt` (it's `undefined`); mempalace hits always do (ISO string). Putting `occurredAt` in the dedupe key would defeat cross-source dedupe entirely — every gbrain/mempalace pair with the same summary would survive because `undefined !== 'YYYY-...'`. Updated assistant.ts:152-157 to explain this.

File: packages/memory-hybrid/src/hybrid-port.ts [hunk 10]
```diff
@@ -174,35 +240,51 @@ export class HybridMemoryPort implements MemoryPort {
...
-   *   1. Explicit routing override for the method name.
-   *   2. Default routing table.
+   * Priority (in order):
+   *   1. Explicit routing override for the method name. If the override names
+   *      the primary but the primary lacks the capability, fall through to
+   *      the secondary anyway — never silently route to a port that can't
+   *      serve the call.
+   *   2. Default routing table preference.
+   *   3. Final fallback: whichever port declares the capability.
    *
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply packages/memory-hybrid/src/hybrid-port.ts]: Addressed in 22e5689 by rewriting the docstring to match the code. The original 3-step description was aspirational — steps 1 and 2 were the same thing (the routing table IS the override), and step 3's capability fallback is already what step 2 does (if the preferred port lacks the capability we fall through). New docstring at hybrid-port.ts:240-256 describes the actual single-rule + capability-fallback logic.

File: .claude/scheduled_tasks.lock [hunk 11]
```diff
@@ -1 +1 @@
-{"sessionId":"f8bc64fc-9881-4489-9cbf-496a51fce7c6","pid":72866,"procStart":"Sat May  9 02:11:52 2026","acquiredAt":1778298894623}
\ No newline at end of file
+{"sessionId":"9208cd9e-7b9c-4bb4-99c2-65ad7f2ee4c5","pid":88031,"procStart":"Mon May 11 03:02:19 2026","acquiredAt":1778470430217}
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply .claude/scheduled_tasks.lock]: Addressed in 22e5689. `git rm --cached .claude/scheduled_tasks.lock` to untrack it, plus added `\.claude/scheduled_tasks.lock` to .gitignore so future sessions don't re-stage it. The file remains on disk locally — only the git tracking is removed. Technically this is a removal-from-main but is the right long-term shape for a per-session PID lock.

File: packages/memory-gbrain-crdb-adapter/src/repository.ts [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+}
+
+function parsePageRow(row: BrainPageRow): BrainPageRow {
+  // pg returns FLOAT8[] as a string like "{0.1,0.2}" only if no parser is
+  // installed; with the default parser it's already number[] — but defensive
+  // parsing is cheap.
+  if (typeof row.embedding === 'string') {
+    row = { ...row, embedding: parsePgArray(row.embedding) };
+  }
+  return {
```
[the remark this answers is not in the document]
Comment (jayzalowitz) [reply packages/memory-gbrain-crdb-adapter/src/repository.ts]: Addressed in 22e5689. Introduced a `RawBrainPageRow` type at repository.ts (`embedding: number[] | string | null`) for the raw pg row shape; `parsePageRow` now takes `RawBrainPageRow` and returns `BrainPageRow` with the narrowed `number[] | null` embedding. The `typeof === 'string'` branch is now type-valid, downstream consumers see the parsed shape, and there's no widening to the public `BrainPageRow` type.
